src/fetch-data.js

```diff
@@ -0,0 +1,197 @@
+// @ts-check
+
+/**
+ * @typedef {Object} BaseNote
+ * @property {string} id - The unique identifier for the note.
+ * @property {string} title - The title of the note.
+ */
+
+/**
+ * @typedef {Object} _NoteWithDescription
+ * @property {string} description - The description of the note.
+ * @property {"details"} type
+ * @typedef {BaseNote & _NoteWithDescription} NoteWithDescription
+ */
+
+/**
+ * @typedef {Object} _NoteWithTasks
+ * @property {string[]} tasks - A list of tasks.
+ * @property {"tasks"} type
+ * @typedef {BaseNote & _NoteWithTasks} NoteWithTasks
+ */
+
+/**
+ * @typedef {NoteWithDescription | NoteWithTasks} Note
+ */
+
+class NotesDataAPI {
+  /**
+   * @type {Map<string, Note>}
+   */
+  notes = new Map();
+
+  /**
+   * Fetch the notes data from the API
+   * @param {string} [url]
+   */
+  fetchData(url) {
+    if (url == undefined) return this.updateNotes(mockData);
+
+    fetch(url)
+      .then((res) => res.json())
+      .then((notes) => this.updateNotes(notes))
+      .catch((e) => console.error(e));
+  }
+
+  /**
+   * Update the list of notes based on their id
+   * @param {readonly Note[]} notes
+   */
+  updateNotes(notes) {
+    notes.forEach((note) => this.updateNote(note, false));
+    this.updatePage();
+  }
+
+  /**
+   * @param {Note} note
+   * @param {boolean} updatePage
+   */
+  updateNote(note, updatePage = true) {
+    const noteData = this.notes.get(note.id) ?? {};
+
+    // merge contents of both objects
+    // might be a bad practice
+    // Silviu help
+    this.notes.set(note.id, { ...noteData, ...note });
+
+    if (updatePage) this.updatePage();
+  }
+
+  updatePage() {
+    const notesConentArea = document.querySelector("#notes-content-area");
+
+    if (notesConentArea == null)
+      return console.error("Missing notes content area");
+
+    notesConentArea.innerHTML = "";
+
+    for (const note of this.notes.values())
+      if (note.type == "details")
+        notesConentArea.appendChild(NoteWithDescriptionHTML(note));
+      else notesConentArea.appendChild(NoteWithTasksHTML(note));
+  }
+}
+
+// TODO: rethink the functions below after we have the markup for a note
+
+/**
+ * Generate the HTML code for a note with details
+ * @param {NoteWithDescription} note
+ * @returns {Node}
+ */
+function NoteWithDescriptionHTML(note) {
```

Comment:
Look into using https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template, I may or may not remember if this is the correct usage for reusable HTML.

Reply:
As a not this was nice of you to add but a bit beyond the scope of the ticket.